packages/portfolio-contract/src/pos-gmp.flows.ts

```diff
@@ -0,0 +1,447 @@
+import type { Amount } from '@agoric/ertp';
+import {
+  makeTracer,
+  mustMatch,
+  NonNullish,
+  type TypedPattern,
+} from '@agoric/internal';
+import type { DenomAmount, Orchestrator } from '@agoric/orchestration';
+import {
+  AxelarGMPMessageType,
+  type AxelarGmpOutgoingMemo,
+  type ContractCall,
+} from '@agoric/orchestration/src/axelar-types.js';
+import {
+  buildGMPPayload,
+  gmpAddresses,
+} from '@agoric/orchestration/src/utils/gmp.js';
+import type { AmountKeywordRecord, ZCFSeat } from '@agoric/zoe';
+import { AmountKeywordRecordShape } from '@agoric/zoe/src/typeGuards.js';
+import { assert, throwRedacted as Fail } from '@endo/errors';
+import { M } from '@endo/patterns';
+import type { GuestInterface } from '../../async-flow/src/types.ts';
+import { AxelarChains, type YieldProtocol } from './constants.js';
+import type { PortfolioKit } from './portfolio.exo.ts';
+import {
+  type PortfolioInstanceContext,
+  provideAccountInfo,
+} from './portfolio.flows.ts';
+import type {
+  AxelarChainsMap,
+  OfferArgsFor,
+  OpenPortfolioGive,
+} from './type-guards.ts';
+
+const trace = makeTracer('GMPF');
+const { keys } = Object;
+
+type AxelarChain = keyof typeof AxelarChains; // XXX rename AxelarChains -> AxelarChain
+type BaseGmpArgs = {
+  destinationEVMChain: AxelarChain;
+  keyword: string;
+  amounts: AmountKeywordRecord;
+};
+const GmpCallType = {
+  ContractCall: 1,
+  ContractCallWithToken: 2,
+} as const;
+type GmpCallType = (typeof GmpCallType)[keyof typeof GmpCallType];
+
+export type GmpArgsContractCall = BaseGmpArgs & {
+  destinationAddress: string;
+  type: GmpCallType;
+  contractInvocationData: Array<ContractCall>;
+};
+type GmpArgsTransferAmount = BaseGmpArgs & {
+  transferAmount: bigint;
+};
+type GmpArgsWithdrawAmount = BaseGmpArgs & {
+  withdrawAmount: bigint;
+};
+const ContractCallShape = M.splitRecord({
```

Comment:
Just a minor nitpicky preference: I think this wall of text could use a blank line or two between logical groupings, if you care to do it.